astropy/coordinates/spectral_coordinate.py

```diff
@@ -0,0 +1,842 @@
+
+        self._observer = value
+
+    @property
+    def target(self):
+        """
+        The coordinate frame of the object being observed.
+
+        Returns
+        -------
+        `~astropy.coordinates.BaseCoordinateFrame`
+            The astropy coordinate frame representing the target.
+        """
+        if self._target_specified:
+            return self._target
+
+    @target.setter
+    def target(self, value):
+        if self.target is not None:
+            raise ValueError("target has already been set")
+
+        self._target_specified = value is not None
+
+        value = self._validate_coordinate(value)
+
+        self._target = value
+
+    @property
+    def radial_velocity(self):
+        """
+        Radial velocity of target relative to the observer.
+
+        Returns
+        -------
+        `~astropy.units.Quantity`
+            Radial velocity of target.
+
+        Notes
+        -----
+        This is different from the ``.radial_velocity`` property of a
+        coordinate frame in that this calculates the radial velocity with
+        respect to the *observer*, not the origin of the frame.
+        """
+        return self._calculate_radial_velocity(self._observer, self._target,
+                                               as_scalar=True)
+
+    @property
+    def redshift(self):
+        """
+        Redshift of target relative to observer. Calculated from the radial
+        velocity.
+
+        Returns
+        -------
+        float
+            Redshift of target.
+        """
+        return _velocity_to_redshift(self.radial_velocity)
+
+    @staticmethod
+    def _calculate_radial_velocity(observer, target, as_scalar=False):
+        """
+        Compute the line-of-sight velocity from the observer to the target.
+
+        Parameters
+        ----------
+        observer : `~astropy.coordinates.BaseCoordinateFrame`
+            The frame of the observer.
+        target : `~astropy.coordinates.BaseCoordinateFrame`
+            The frame of the target.
+        as_scalar : bool
+            If `True`, the magnitude of the velocity vector will be returned,
+            otherwise the full vector will be returned.
+
+        Returns
+        -------
+        `~astropy.units.Quantity`
+            The radial velocity of the target with respect to the observer.
+        """
+
+        # Convert observer and target to ICRS to avoid finite differencing
+        # calculations that lack numerical precision.
+        observer_icrs = observer.transform_to(ICRS)
+        target_icrs = target.transform_to(ICRS)
+
+        pos_hat = SpectralCoord._norm_d_pos(observer_icrs, target_icrs)
+
+        d_vel = target_icrs.velocity - observer_icrs.velocity
+
+        vel_mag = np.dot(d_vel.d_xyz, pos_hat.xyz)
+
+        if NUMPY_LT_1_17:
+            vel_mag *= d_vel.d_xyz.unit
+
+        if as_scalar:
+            return vel_mag
+        else:
+            return vel_mag * pos_hat.xyz
```

Comment:
I'd just return `vel_mag * pos_hat` to keep using representations. Saves you trouble later!

Reply:
Done